pandas/core/construction.py

```diff
@@ -450,6 +450,15 @@ def sanitize_array(
         subarr = _try_cast(arr, dtype, copy, raise_cast_failure)
     elif isinstance(data, abc.Set):
         raise TypeError("Set type is unordered")
+    elif lib.is_scalar(data) and index is not None and dtype is not None:
+        try:
+            data = maybe_cast_to_datetime(data, dtype)
+        except TypeError:
+            pass
```

Comment:
Do you know for what case it raises? As I would think that `maybe_cast_to_datetime` is not supposed to raise (or if it raises, it's an error that should bubble up to the user), so that might need a fix there?

Reply:
> Do you know for what case it raises?

it raises for string type with a string that satisfies
```
>>> pd.core.dtypes.common.is_datetime64_dtype("M")
True
```
The scalar used in extension tests is random and is sometimes, say "M", therefore a separate test has been added. 




> As I would think that `maybe_cast_to_datetime` is not supposed to raise (or if it raises, it's an error that should bubble up to the user), so that might need a fix there?

it raises `TypeError: Cannot cast datetime64 to string`

https://github.com/pandas-dev/pandas/blob/1a82659ec66995eacc74e4b39acf648ffbd5aede/pandas/core/dtypes/cast.py#L1387-L1395

I suspect that maybe_cast_to_datetime shouldn't raise in this case, seems to contradict the maybe of maybe_cast_to_datetime

Reply:
I would say that's a bug in `maybe_cast_datetime`. We should *not* check the actual *value* if it is a datetime64 dtype, but only if that value (eg because it being an array) has a dtype. 
(that's related to passing the actual dtype vs the array to the `is_..._dtype` functions, something we have been discussing lately)

Can you see if this fixes it for you:

```patch
--- a/pandas/core/dtypes/cast.py
+++ b/pandas/core/dtypes/cast.py
@@ -1384,7 +1384,7 @@ def maybe_cast_to_datetime(value, dtype, errors: str = "raise"):
                         pass
 
         # coerce datetimelike to object
-        elif is_datetime64_dtype(value) and not is_datetime64_dtype(dtype):
+        elif is_datetime64_dtype(getattr(value, "dtype", None)) and not is_datetime64_dtype(dtype):
             if is_object_dtype(dtype):
                 if value.dtype != DT64NS_DTYPE:
                     value = value.astype(DT64NS_DTYPE)
```

Reply:
> Can you see if this fixes it for you:

thanks.